crates/bevy_ecs/src/query/iter.rs

```diff
@@ -812,3 +1593,159 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIterationCursor<'w, 's, D, F> {
         }
     }
 }
+
+// A wrapper struct that gives its data a neutral ordering.
+#[derive(Copy, Clone)]
+struct NeutralOrd<T>(T);
+
+impl<T> PartialEq for NeutralOrd<T> {
+    fn eq(&self, _other: &Self) -> bool {
+        true
+    }
+}
+
+impl<T> Eq for NeutralOrd<T> {}
+
+#[allow(clippy::non_canonical_partial_ord_impl)]
+impl<T> PartialOrd for NeutralOrd<T> {
+    fn partial_cmp(&self, _other: &Self) -> Option<Ordering> {
+        Some(Ordering::Equal)
+    }
+}
+
+impl<T> Ord for NeutralOrd<T> {
+    fn cmp(&self, _other: &Self) -> Ordering {
+        Ordering::Equal
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I think I would still like to see tests that show correct sorted iteration in the non-panic case.
You mention that `The underlying sorts are just the slice sorts, so correct order does not need to be tested.`
but that's more of an implementation detail; I think it'd be nice to have a simple test showing that it works.

Maybe a test for the `Entity` QueryLens that you mentioned would be nice (which depends on `(Entity, Entity)` being a valid QueryData))

Reply:
added the `query_sorts` test!